Expanded BigInt support

robdmoore opened the pull request:
Took https://github.com/msgpack/msgpack-javascript/pull/211, rebased against latest `main`.

BigInt support had been added (in a very basic way) in msgpack library, these changes expand the int handling to have more configurability and sophistication, while maintaining backwards compatibility with latest msgpack.

I also refactored the tests to pull in a whole heap of external dependencies that need to be here so they are more easily understood and modified (I had to modify them to handle BigInt to avoid tests breaking).

jasonpaulos committed:
feat: Expand bigint support to have more configurability

robdmoore committed:
test: Removing external dependencies for tests so they can be modified
Modified them to properly support bigint and convert them to TypeScript
Adding VS Code support for mocha

robdmoore commented on .vscode/settings.json:
```diff
@@ -2,13 +2,13 @@
   "typescript.tsdk": "node_modules/typescript/lib",
   "files.eol": "\n",
   "editor.tabSize": 2,
+  "editor.defaultFormatter": "esbenp.prettier-vscode",
+  "editor.formatOnSave": true,
```
Prettier was already in there, so I figured it was ok to add this, I was sick of not having it auto-format while making changes.

robdmoore commented on .vscode/settings.json:
```diff
@@ -2,13 +2,13 @@
-    "whatwg"
-  ]
+  "cSpell.words": ["instanceof", "tsdoc", "typeof", "whatwg"],
+  "mochaExplorer.files": "test/**/*.test.{ts,js}",
```
Added mochaexplorer support so it was easier to run tests in VS Code.

robdmoore commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# MessagePack for JavaScript/ECMA-262  <!-- omit in toc -->
+# MessagePack for JavaScript/ECMA-262 <!-- omit in toc -->
```
FYI Some prettier formatting changes in this file along with some real changes too

robdmoore commented on README.md:
```diff
@@ -143,19 +145,32 @@ NodeJS `Buffer` is also acceptable because it is a subclass of `Uint8Array`.
+- `IntMode.UNSAFE_NUMBER`: Always returns the value as a number. Be aware that there will be a loss of precision if the value is outside the range of `Number.MIN_SAFE_INTEGER` to `Number.MAX_SAFE_INTEGER`.
+- `IntMode.SAFE_NUMBER`: Always returns the value as a number, but throws an error if the value is outside of the range of `Number.MIN_SAFE_INTEGER` to `Number.MAX_SAFE_INTEGER`.
+- `IntMode.MIXED`: Returns all values inside the range of `Number.MIN_SAFE_INTEGER` to `Number.MAX_SAFE_INTEGER` as numbers and all values outside that range as bigints.
+- `IntMode.AS_ENCODED`: Returns all values encoded as int64/uint64 as bigints and all other integers as numbers.
```
I added this new int mode, which matched the BigInt support that was added to the main library.

robdmoore replied:
(and felt like a useful mode too)

jasonpaulos replied:
Seems useful, good idea

robdmoore commented on package.json:
```diff
@@ -73,7 +73,9 @@
     "lodash": "latest",
     "mocha": "latest",
-    "msgpack-test-js": "latest",
+    "msg-ext": "^1.0.1",
+    "msg-int64": "^0.1.1",
+    "msg-timestamp": "^1.0.1",
```
These 3 were deep dependencies of `msgpack-test-js`, the rest of which I inlined into this project (see the new `test/utils` folder).

jasonpaulos replied:
Can you explain why you had to inline `msgpack-test-js`? Was something broken that you had to fix -- if so, what specifically did you change?

robdmoore replied:
Yeah, the core of it was JSON.stringify'ing an object that had BigInt in it and dying

robdmoore replied:
Plus frankly, it was weird that stuff was external anyway, it's code that clearly should live in this project.

robdmoore commented on src/Encoder.ts:
```diff
@@ -200,7 +213,7 @@ export class Encoder<ContextType = undefined> {
           this.writeU8(0xce);
           this.writeU32(object);
-        } else if (!this.useBigInt64) {
+        } else if (this.useInt64) {
```
The `useBigInt64` option was being conflated for 2 things - one was encoding bigint's as int64, the other was encoding big `number`'s as `int64` rather than `float64` so I've added configuration to support both (`useInt64` and `forceBigIntToInt64`). The default config (both `false`) will have the same behaviour as before.

robdmoore replied:
Actually, on reflection, maybe the default behaviour here is different and I need to make `useInt64` true by default?

jasonpaulos replied:
Here's my opinion. The upstream library's decision to encode numbers above the uint32 max or below the int32 min if `useBigInt64` is false is nonsensical. Javascript numbers can safely represent integers outside of this range without using bigints, and the decision to encode these as floats when using the default encoding params (thereby confusing every other non-JS msgp decoder) is mind boggling.

I'd get rid of this condition entirely.

At this point our top priority is delivering a library with enhanced features that are easy to use and make sense. We're only in this position because the upstream library has failed us. Let's focus on correcting the behavior over trying to be compatible with it.

robdmoore replied:
I'm happy to drop compatibility with the core library if you are. My main concern is that it means we do have to maintain this library long term since it makes it less likely that the core library would accept our changes if they take on maintenance of that library again.

Side-note: should we raise all of these PRs to the core library as well as this one do you think or not worth it?

robdmoore replied:
Also, the fact there is a forceIntegerToFloat means that it's still possible to force a BigInt to a float.

jasonpaulos replied:
> Also, the fact there is a forceIntegerToFloat means that it's still possible to force a BigInt to a float.

We should fix this inconsistency, since the docstring for `forceBigIntToInt64` says `forceIntegerToFloat` doesn't affect bigints. I'm fine with qualifying that dosctring to say that it doesn't affect bigints if that option is enabled.

> should we raise all of these PRs to the core library as well as this one do you think or not worth it?

For the most part each of these PRs already exists upstream, right? I guess the main difference is that they're not based on the most recent changes to the main branch.

My current intention is to release this fork under a different name, then post a link to it on each of the upstream PRs. Hopefully then we'll at least help others in similar situations.

Ideally yes, we could merge all of these features back upstream, but I'm not counting on that as a viable option in the short term.

jasonpaulos commented:
Closing and reopening to trigger github actions

jasonpaulos commented on src/Decoder.ts:
```diff
@@ -371,11 +378,11 @@ export class Decoder<ContextType = undefined> {
         if (headByte < 0x80) {
           // positive fixint (0xxx xxxx) 0x00 - 0x7f
-          object = headByte;
+          object = this.convertNumber(headByte);
```
Good catch, this wasn't in the original PR

jasonpaulos commented on src/Encoder.ts:
```diff
@@ -1,9 +1,9 @@
 import { ensureUint8Array } from "./utils/typedArrays";
 import type { ExtData } from "./ExtData";
 import type { ContextOf } from "./context";
+import { setInt64, setUint64 } from "./utils/int";
```
nit: why reorder import? I think some linters complain about normal imports after type-only imports

robdmoore replied:
Prettier did that automatically I believe.

robdmoore replied:
Oh wait, no. I think it was because I originally removed it and then realised it was actually needed. I'll fix this.

jasonpaulos commented on src/Encoder.ts:
```diff
@@ -144,8 +157,8 @@ export class Encoder<ContextType = undefined> {
       this.encodeString(object);
-    } else if (this.useBigInt64 && typeof object === "bigint") {
-      this.encodeBigInt64(object);
+    } else if (this.forceBigIntToInt64 && typeof object === "bigint") {
```
Here's my proposal for encoding options: let's keep `useBigInt64`, but make it default to true. It would only be used here (i.e. `} else if (this.useBigInt64 && typeof object === "bigint") {`).

The only reason to ever disable it is if you have a custom extension for encoding bigints from a legacy msgp version.

That way we can remove the changes to `encodeObject` and the route for encoding a bigint is simpler.

robdmoore replied:
I still think it's useful to allow `forceBigIntToInt64` option since you then have an ability to ensure a fixed length encoding by using bigints. It feels like a really useful mechanism. I've simplified the logic based on your suggestion though, but kept backwards compatibility with custom extensions because it didn't negatively impact things.

jasonpaulos replied:
Ok, that seems reasonable to me

jasonpaulos commented on test/bigint64.test.ts:
```diff
@@ -1,5 +1,6 @@
 import assert from "assert";
-import { encode, decode } from "../src";
+import { encode, decode, IntMode, ExtensionCodec, DecodeError } from "../src";
+import { getInt64, getUint64 } from "src/utils/int";
```
```suggestion
import { getInt64, getUint64 } from "../src/utils/int";
```

I believe the CI browser tests are complaining about this

robdmoore committed:
Review feedback
BREAKING CHANGE: `useBigInt64` deprecated from Encoder and > 32 bit ints now get encoded as a uint64 rather than float64

robdmoore commented:
FYI I pushed another commit with the tweaks based on feedback

robdmoore commented on src/Encoder.ts:
```diff
@@ -150,15 +138,9 @@ export class Encoder<ContextType = undefined> {
-      if (!this.forceIntegerToFloat) {
-        this.encodeNumber(object);
-      } else {
-        this.encodeNumberAsFloat(object);
```
This was already in encodeNumber

jasonpaulos commented on test/bigint64.test.ts:
```diff
@@ -10,29 +11,345 @@ describe("useBigInt64: true", () => {
+}
+
+// declared as a function to delay referencing the BigInt constructor
+function BIGINTSPECS(): Record<string, TestCase> {
```
It seems that there's a lot of duplicate testing here and in `codec-bigint.test.ts` -- if these are in some way different, let me know. Otherwise I'd suggest removing the duplicate tests/helpers from this file. They seem to fit better in `codec-bigint.test.ts`

jasonpaulos commented on test/codec-bigint.test.ts:
It seems like this file wasn't formatted with prettier, since I'm seeing local changes upon saving

jasonpaulos commented on test/encode.test.ts:
It would be nice to see some tests of `forceBigIntToInt64` here

I'd also like to see tests of decoding with `IntMode.AS_ENCODED` -- as of right now I'm not sure if this is tested anywhere

jasonpaulos reviewed (commented):
Implementation is looking good (let's just address the `forceIntegerToFloat` issue).

I left a few test comments as well.

jasonpaulos committed:
Additional int testing

robdmoore committed:
Merge pull request #3 from jasonpaulos/additional-int-testing

Additional int testing

jasonpaulos commented on src/Encoder.ts:
```diff
@@ -13,14 +13,14 @@ export type EncoderOptions<ContextType = undefined> = Partial<
     /**
-     * Encodes bigint as Int64 or Uint64 if it's set to true.
+     * Encodes bigint as Int64 or Uint64 if it's set to true, regardless of the size of bigint number.
      * {@link forceIntegerToFloat} does not affect bigint.
```
```suggestion
     * {@link forceIntegerToFloat} does not affect bigint if this is enabled.
```

This is my suggestion to fix the inconsistency. This looks good to merge after this is addresses

jasonpaulos commented on README.md:
```diff
@@ -44,15 +44,16 @@ deepStrictEqual(decode(encoded), object);
-    - [Handling BigInt with ExtensionCodec](#handling-bigint-with-extensioncodec)
-    - [The temporal module as timestamp extensions](#the-temporal-module-as-timestamp-extensions)
+  - [ExtensionCodec context](#extensioncodec-context)
+  - [Handling BigInt and big numbers](#handling-bigint-and-big-numbers)
```
nit: I think just "Handling BigInt" would be sufficient

jasonpaulos commented on README.md:
```diff
@@ -352,19 +368,28 @@ const encoded = = encode({myType: new MyType<any>()}, { extensionCodec, context
+**Decoding**
 
-This library does not handle BigInt by default, but you have two options to handle it:
+This library does not handle decoding BigInt by default, but you have three options to handle it:
```
Actually we default to `IntMode.AS_ENCODED` if it's not specified, so this is no longer accurate

robdmoore replied:
Default is actually `UNSAFE_NUMBER`

jasonpaulos replied:
Good point, I read it wrong

robdmoore committed:
docs: Documentation improvements based on PR feedback

robdmoore commented:
@jasonpaulos Ready for final review

jasonpaulos reviewed (approved):
Looks good!
